### ISSUE TITLE
misdetection of spinloop in GenMC v0.6

### ISSUE THREAD

**Issue by db7** [MPI-SWS/genmc#15]
We have a few examples where GenMC v0.6 cannot properly detect the spinloops.  Below is a simple MCS lock test case using stdatomic with explicit notation.

Depending on how the following loop is implemented, genmc hangs or not:

```c
#ifdef EXPLICIT
	while (atomic_load_explicit(&me->next, memory_order_seq_cst) == NULL);
#else
	while (me->next == NULL);
#endif
```

```
genmc -- -DEXPLICIT mcs-sc.c ## hangs
genmc -- mcs-sc.c ## terminates
```

Note that they are equivalent wrt to barrier mode (`me->next` is an atomic pointer). Also, whether or not I am using `-mo`, `-imm`, `-check-liveness` is irrelevant, it hangs as well. The same problem occurs if we are using __atomic builtins. 

Is there a way how to help genmc to know what to monitor as spinloop?

Here is the test case:
```c
#include <pthread.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdint.h>
#include <assert.h>
#include <stdatomic.h>
/*******************************************************************************
 * MCS lock implementation
 ******************************************************************************/
typedef struct mcs_node_s {
	_Atomic(struct mcs_node_s*) next;
	atomic_int spin;
} mcs_node_t;

typedef struct {
	_Atomic(struct mcs_node_s*) tail;
} mcs_lock_t;

static inline void
mcslock_acquire(mcs_lock_t *m, mcs_node_t *me)
{
	mcs_node_t *tail;

	atomic_store_explicit(&me->next, NULL, memory_order_seq_cst);
	atomic_store_explicit(&me->spin, 1, memory_order_seq_cst);

	tail = atomic_exchange_explicit(&m->tail, me, memory_order_seq_cst);
	if (!tail)
		return;

	atomic_store_explicit(&tail->next, me, memory_order_seq_cst);
	while (atomic_load_explicit(&me->spin, memory_order_seq_cst));
}

static inline void
mcslock_release(mcs_lock_t *m, mcs_node_t *me)
{
	if (!atomic_load_explicit(&me->next, memory_order_seq_cst)) {
		struct mcs_node_s* old = me;
		if (atomic_compare_exchange_strong_explicit(
					&m->tail, &old, NULL,
					memory_order_seq_cst,
					memory_order_seq_cst)) {
			return;
		}
#ifdef EXPLICIT
		while (atomic_load_explicit(&me->next, memory_order_seq_cst) == NULL);
#else
		while (me->next == NULL);
#endif
	}
	struct mcs_node_s * next = atomic_load_explicit(&me->next,
			memory_order_seq_cst);
	atomic_store_explicit(&next->spin, 0, memory_order_seq_cst);
}
/*******************************************************************************
 * test
 ******************************************************************************/
#define NTHREADS 3
static mcs_lock_t __lock;
static mcs_node_t __nodes[NTHREADS];
static unsigned int x;
static void* run(void *arg) {
	uint32_t cpu = (intptr_t) arg;
	mcslock_acquire(&__lock, &__nodes[cpu]);
	x++;
	mcslock_release(&__lock, &__nodes[cpu]);
	return NULL;
}
int main() {
	pthread_t t[NTHREADS];
	for (intptr_t i = 0; i < NTHREADS; i++) pthread_create(t+i, 0, run, (void*)i);
	for (intptr_t i = 0; i < NTHREADS; i++) pthread_join(t[i], NULL);
	assert(x == NTHREADS && "unexpected sum");
	return 0;
}
```

**Comment by db7** [comment]
Note that the other spinloo, ie, `while (atomic_load_explicit(&me->spin, memory_order_seq_cst))`, seem to be detected properly.

**Comment by michaliskok** [comment]
Thanks for the report.

The problem is that clang generates a different LLVM-IR code for these two cases: in the case where `-DEXPLICIT` is used, it generates a store to a temporary variable within the loop, thereby preventing it from being recognized as a spinloop by GenMC.

I will try to make GenMC aware of such cases, but this might take some time, as I have a huge backlog at the moment. Possible workarounds include manually transforming the loop into an `assume()` statement, or using a `__VERIFIER_spin_end(0)` at the end of the loop body (also see #16).

**Comment by JonasOberhauser** [comment]
> I will try to make GenMC aware of such cases, but this might take some time, as I have a huge backlog at the moment. Possible workarounds include manually transforming the loop into an `assume()` statement, or using a `__VERIFIER_spin_end(0)` at the end of the loop body (also see #16).

Note that stores to temporaries in general are not a problem, provided they aren't read by other threads or (by this thread) after this loop iteration. 

Maybe the second part can be checked statically and for the first part a dynamic check might suffice --- if another thread reads from a store in a loop that is blocking the thread, unblock the thread. Maybe one can show that this suffices to generate all relevant executions?

**Comment by michaliskok** [comment]
Good point. In fact, GenMC already does something similar to what you are suggesting. The problem is that GenMC's optimization is not generally sound if there are release (or stronger) accesses between the temporary allocation and the store to the temporary, and this is why it gives up in this case (i.e., due to the SC accesses).

Optimizations aside, I think it would be beneficial to get rid of such temporaries in general.

**Comment by michaliskok** [comment]
Fixed in v0.6.1. There is a small chance that GenMC removes temporaries more aggressively than it should. If you notice any weird behavior/missed executions try `-disable-cast-elimination`.
